standard/expressions.md

```diff
@@ -2221,29 +2218,91 @@ The *primary_expression* of an *element_access* shall not be an *array_creation_
 
-The run-time processing of an array access of the form `P[A]`, where `P` is a *primary_expression* of an *array_type* and `A` is an *argument_list*, consists of the following steps:
+The run-time processing of an array access of the form `P[A]`, where `P` is a *primary_expression* of an *array_type* and `A` is an *argument_list* of index expressions, consists of the following steps:
 
 - `P` is evaluated. If this evaluation causes an exception, no further steps are executed.
-- The index expressions of the *argument_list* are evaluated in order, from left to right. Following evaluation of each index expression, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) to one of the following types is performed: `int`, `uint`, `long`, `ulong`. The first type in this list for which an implicit conversion exists is chosen. For instance, if the index expression is of type `short` then an implicit conversion to `int` is performed, since implicit conversions from `short` to `int` and from `short` to `long` are possible. If evaluation of an index expression or the subsequent implicit conversion causes an exception, then no further index expressions are evaluated and no further steps are executed.
+- For each index expression in the *argument_list* in order, from left to right:
+  - The index expression is evaluated, let the type of the resultant value be *T*;
+  - This value is then converted to the first of the types: `int`, `uint`, `long`, `ulong`, or for static bound single rank array access only `Index` or `Range`; for which an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) from *T* exists.
+  - If evaluation of an index expression or the subsequent implicit conversion causes an exception, then no further index expressions are evaluated and no further steps are executed.
 - The value of `P` is checked to be valid. If the value of `P` is `null`, a `System.NullReferenceException` is thrown and no further steps are executed.
-- The value of each expression in the *argument_list* is checked against the actual bounds of each dimension of the array instance referenced by `P`. If one or more values are out of range, a `System.IndexOutOfRangeException` is thrown and no further steps are executed.
-- The location of the array element given by the index expressions is computed, and this location becomes the result of the array access.
+- If the preceding steps have produced a single index value of type `Range` then:
+  - The result of evaluating the array access is a new array.
+  - The starting and ending indices of the `Range` value are determined with respect to the length of the array referenced by `P`.
+  - If either index is out of range of the bounds of the array referenced by `P`, or the ending index comes before the starting index, then a `System.ArgumentOutOfRangeException` is thrown and no further steps are executed.
```

Comment:
Per conversation at the committee, this needs a bit of work. The following is valid:

```csharp
        int[] array = { 0, 1, 2, 3, 4, 5, 6, 7, 8 };
        int[] empty = { };

        var slice1 = array[0..0];
        Console.WriteLine(slice1.Length);
        var slice2 = empty[0..0];
        Console.WriteLine(slice2.Length);
```

Reply:
Addressed in next commit